Add workspace support for the tracking store

mprahl opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/mprahl/mlflow/pull/19116?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/19116/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/19116/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/19116/merge
```

</p>
</details>

<!-- Remove unused checkboxes except for the patch release section at the bottom -->

### Related Issues/PRs

Third one from https://github.com/mlflow/mlflow/pull/18909.

#1464
#5844
Design document: https://docs.google.com/document/d/1IbFfceCmWV3knJfc0ninn58EXLVbHUh1meV_pknOM40/edit

### What changes are proposed in this pull request?

Add a workspace aware tracking SqlAlchemyStore implementation that is
conditionally turned on when workspaces are enabled. The base tracking
store had some minor modifications, mostly guards to ensure things
referenced by ID as method inputs existed. The workspace aware store
further filters this by the workspace column too.

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/prompts`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- Do not modify or remove any text inside the parentheses. Keep both checkboxes below. -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

B-Step62 commented on mlflow/entities/evaluation_dataset.py:
```diff
@@ -322,7 +333,15 @@ def to_df(self) -> "pd.DataFrame":
         import pandas as pd
 
-        records = self.records
+        try:
```
Is this change related to this PR?

mprahl replied:
It was an issue I encountered during the tests but I don't recall if it was directly due to the workspace aware changes.

B-Step62 commented on mlflow/server/handlers.py:
```diff
@@ -2804,6 +2825,54 @@ def _test_webhook(webhook_id: str):
+    if not MLFLOW_ENABLE_WORKSPACES.get():
+        return artifact_path
+
+    workspace = workspace_context.get_current_workspace()
```
This method is called at the server side, does `get_current_workspace` work? It means the server only handles single workspace, which does not satisfy the core requirement.

B-Step62 commented on mlflow/server/handlers.py:
```diff
@@ -2804,6 +2825,54 @@ def _test_webhook(webhook_id: str):
+        if requested_workspace != workspace:
+            raise MlflowException.invalid_parameter_value(
+                f"Artifact path targets workspace '{requested_workspace}' "
+                f"but active workspace is '{workspace}'."
```
```suggestion
                f"but the workspace specified in the request is '{workspace}'."
```

nit: When this message is surfaced at client side, "active" (which is server side concept) might not be obvious.

mprahl commented on mlflow/store/tracking/rest_store.py:
```diff
@@ -155,6 +155,43 @@ class RestStore(AbstractStore):
+            )
+        except Exception as exc:  # pragma: no cover - network errors vary
+            _logger.debug("Failed to probe workspace support: %s", exc)
+            return False
```
Good catch!

ask664 commented:
@B-Step62 @mprahl I added some review comments on the PR, but they’re still showing in a pending state. I’m not sure whether you’re able to see them yet.

ask664 commented:
> @ask664 I cannot see them, did you hit "submit review"?

Ahh, my mistake — I forgot to hit “Submit review.”

B-Step62 commented on mlflow/entities/evaluation_dataset.py:
```diff
@@ -416,6 +435,7 @@ def to_dict(self) -> dict[str, Any]:
                 "created_by": self.created_by,
                 "last_updated_by": self.last_updated_by,
                 "experiment_ids": self.experiment_ids,
+                "workspace": self.workspace,
```
> so clients can correctly identify the tenant context of each experiment

@ask664 UI will always have workspace in the URL and only show entities under the workspace scope. For SDK, I believe we should always ask users to set active workspace through environment variable or `set_workspace` explicitly, rather than inferring from the object metadata, such that it is guaranteed a client python process never deal with multiple workspaces at the same time. Do you have any use case that cannot be handled by the env var?

ask664 replied:
Thanks for the clarification. Yes, the UI URL and the SDK’s active workspace setting will cover most cases.

mprahl commented on mlflow/server/workspace_helpers.py:
```diff
@@ -48,4 +71,53 @@ def _get_workspace_store(workspace_uri: str | None = None, tracking_uri: str | N
+    if not MLFLOW_ENABLE_WORKSPACES.get():
+        header_workspace = request.headers.get(WORKSPACE_HEADER_NAME)
+        header_workspace = header_workspace.strip() if header_workspace else None
+        if header_workspace:
```
Done

mprahl commented on mlflow/store/tracking/rest_store.py:
```diff
@@ -169,7 +206,7 @@ def _get_server_version(host_creds: MlflowHostCreds) -> Version | None:
         """
         try:
-            response = http_request(
+            response = rest_utils.http_request(
```
Done

mprahl committed:
Add workspace support to SQLAlchemy tracking store

Add a workspace aware tracking SqlAlchemyStore implementation that is
conditionally turned on when workspaces are enabled. The base tracking
store had some minor modifications, mostly guards to ensure things
referenced by ID as method inputs existed. The workspace aware store
further filters this by the workspace column too.

Signed-off-by: mprahl <mprahl@users.noreply.github.com>

BenWilson2 commented on mlflow/store/tracking/sqlalchemy_store.py:
```diff
@@ -404,13 +457,15 @@ def compute_next_token(current_size):
 
+            experiment_filters = [
+                *attribute_filters,
+                SqlExperiment.lifecycle_stage.in_(lifecycle_stags),
```
`lifecycle_stages`? Can we just fix this typo since we're poking around in these variables anyway?

BenWilson2 commented on mlflow/store/tracking/sqlalchemy_store.py:
```diff
@@ -2681,6 +2810,11 @@ def search_traces(
         with self.ManagedSessionMaker() as session:
+            # Filter locations (experiment_ids) to only include accessible experiments
+            locations = self._filter_experiment_ids(session, locations)
+            if not locations:
```
similar comment here - do we need the explicit return like this?

BenWilson2 commented on mlflow/store/tracking/sqlalchemy_store.py:
```diff
@@ -4744,6 +4933,364 @@ def _get_orderby_clauses(order_by_list, session):
+                error_code=INVALID_STATE,
+            )
+
+        reserved_prefix = append_to_uri_path(self.artifact_root_uri, "workspaces").rstrip("/") + "/"
```
We shouldn't be string munging unix paths. This definitely won't work on NTFS. Use pathlib.

BenWilson2 commented on mlflow/utils/workspace_utils.py:
```diff
@@ -1,5 +1,94 @@
-"""Workspace-related utility constants."""
+"""Workspace-related utility constants and helpers."""
```
I think we can get rid of the module-level docstring. The name says it all. 

The rest of the docstrings in this file could benefit from standard Google style docstrings (args, returns), though.

mprahl commented on mlflow/store/tracking/sqlalchemy_store.py:
```diff
@@ -1560,6 +1690,9 @@ def compute_next_token(current_size):
             offset = SearchUtils.parse_start_offset_from_page_token(page_token)
             experiment_ids = [int(e) for e in experiment_ids]
+            experiment_ids = self._filter_experiment_ids(session, experiment_ids)
+            if not experiment_ids:
```
Good point! It seems like `._in` below handles this well.

mprahl commented on mlflow/store/tracking/sqlalchemy_store.py:
```diff
@@ -2801,6 +2937,8 @@ def delete_trace_tag(self, trace_id: str, key: str):
         """
         with self.ManagedSessionMaker() as session:
+            # Validate trace exists and is accessible
+            self._get_trace_record(session, trace_id)
```
In single-tenant mode, it's not needed. In `WorkspaceAwareSqlAlchemyStore`, this adds the additional workspace filter to ensure the `trace_id` is bound to the workspace. If not, it raises a `RESOURCE_DOES_NOT_EXIST` error.

mprahl commented on mlflow/store/tracking/sqlalchemy_store.py:
```diff
@@ -3118,6 +3249,16 @@ def link_traces_to_run(self, trace_ids: list[str], run_id: str) -> None:
 
         with self.ManagedSessionMaker() as session:
+            # Validate run exists and is accessible
+            self._get_run(session, run_id)
```
In this case, the error would be "Run with id={run_uuid} not found", which I think is a helpful error message in case you put the wrong run ID.

In terms of the single tenant mode, it does add an additional database query. It's the trade off of the SqlAlchemyStory being the base class that isn't workspace aware. My original implementation was adding workspace specific filtering only when workspaces were enabled in SqlAlchemyStore but @B-Step62 pointed out that'd be harder to maintain.

Perhaps a compromise is to have helper methods of `_get_run_exists` or similar that are cheaper as they'd just hit the database index and would only return a boolean. What are your thouhts @BenWilson2 and @B-Step62 ?

mprahl commented on mlflow/store/tracking/sqlalchemy_store.py:
```diff
@@ -4137,6 +4293,14 @@ def _load_dataset_records(
         with self.ManagedSessionMaker() as session:
+            dataset = self._dataset_query(session).filter_by(dataset_id=dataset_id).first()
+
+            if not dataset:
```
This one is also necessary because we want to error out when the dataset by that ID isn't in the workspace. In single tenant mode, it's not really needed but keeps the code consistent.

mprahl committed:
Make the default SQLite tracking URI use absolute paths

This helps for cases where the default tracking URI is used multiple
times but in different working directories to avoid the wrong
cached value being returned from _get_store. This can happen in the
tests for example.

Signed-off-by: mprahl <mprahl@users.noreply.github.com>

BenWilson2 committed:
Fix R XGBoost tests (#19227)

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

B-Step62 commented on mlflow/entities/experiment.py:
```diff
@@ -75,6 +78,11 @@ def last_update_time(self):
         self._last_update_time = last_update_time
 
+    @property
+    def workspace(self):
```
```suggestion
    def workspace(self) -> str | None:
```

B-Step62 commented on mlflow/store/tracking/sqlalchemy_store.py:
```diff
@@ -259,11 +263,68 @@ def __init__(self, db_uri, default_artifact_root):
+        path = parsed.path if parsed.scheme else uri
+        return [segment for segment in path.split("/") if segment]
+
+    def _add_fields(self, instance):
```
```suggestion
    def _with_workspace_field(self, instance):
```
nit: since we don't plan to add more fields in near future, an explicit name would be easier to reason about.

B-Step62 commented on mlflow/store/tracking/sqlalchemy_store.py:
```diff
@@ -353,22 +415,27 @@ def create_experiment(self, name, artifact_location=None, tags=None):
+                    experiment.artifact_location = self._get_artifact_location(
+                        experiment.experiment_id
+                    )
+                    session.flush()
```
The `ManagedSessionMaker` will commit the change when existing the context manager, so I think we don't need this second flush.

B-Step62 commented on mlflow/utils/file_utils.py:
```diff
@@ -506,7 +506,8 @@ def local_file_uri_to_path(uri):
         if is_windows() and parsed_path.netloc:
             return urllib.request.url2pathname(rf"\\{parsed_path.netloc}{path}")
-    return urllib.request.url2pathname(path)
+    converted = urllib.request.url2pathname(path)
```
What is this change for?

B-Step62 commented on mlflow/utils/workspace_utils.py:
```diff
@@ -1,5 +1,122 @@
+    candidate = _normalize_workspace(workspace)
+    if candidate:
+        return candidate
+
```
```suggestion
    if candidate := _normalize_workspace(workspace)
        return candidate

```

B-Step62 commented on tests/entities/test_evaluation_dataset.py:
```diff
@@ -300,8 +300,19 @@ def test_evaluation_dataset_complex_tags():
 
 
-def test_evaluation_dataset_to_df():
+def test_evaluation_dataset_to_df(monkeypatch):
```
What is this patching for?

mprahl commented on mlflow/store/tracking/sqlalchemy_store.py:
```diff
@@ -629,6 +768,23 @@ def _get_run(self, session, run_uuid, eager=False):
+        del for_update_or_delete
+        return self._get_query(session, SqlTraceInfo)
+
+    def _get_trace_record(self, session, trace_id: str) -> SqlTraceInfo:
```
Good catch, they aren't used after the refactor. I'll remove it!

mprahl commented on mlflow/store/workspace_aware_mixin.py:
```diff
@@ -0,0 +1,65 @@
+        )
+
+    @contextmanager
+    def _workspace_session(self):
```
Good catch. I missed that in the refactor.

B-Step62 reviewed (approved):
LGTM!
